signer/migrations/0011__materialized_chains.sql

```diff
@@ -0,0 +1,148 @@
+-- used in queries, the `bitcoin_chain_tip` column should be used to filter the
+-- results to the specific chain tip which you are interested in. The results
+-- are not ordered by block height, so you should do that yourself if you
+-- require the results to be ordered.
+CREATE VIEW sbtc_signer.canonical_bitcoin_chain_view AS
+SELECT DISTINCT
+    bitcoin_chain_tip,
+    bitcoin_block_hash,
+    bitcoin_block_height
+FROM sbtc_signer.canonical_chains;
+
+-- Indexes to support common queries on the canonical chains table.
+CREATE INDEX ix_canonical_chains_run_id ON sbtc_signer.canonical_chains(run_id);
+CREATE INDEX ix_canonical_chains_bitcoin_chain_tip ON sbtc_signer.canonical_chains(bitcoin_chain_tip);
+CREATE INDEX ix_canonical_chains_bitcoin_block_hash ON sbtc_signer.canonical_chains(bitcoin_chain_tip, bitcoin_block_hash);
+CREATE INDEX ix_canonical_chains_bitcoin_block_height ON sbtc_signer.canonical_chains(bitcoin_chain_tip, bitcoin_block_height);
+CREATE INDEX ix_canonical_chains_stacks_block_hash ON sbtc_signer.canonical_chains(bitcoin_chain_tip, stacks_block_hash);
+CREATE INDEX ix_canonical_chains_stacks_block_height ON sbtc_signer.canonical_chains(bitcoin_chain_tip, stacks_block_height);
+
+-- New sequence which will be used to generate the run_id for each materialized
+-- view canonical chain.
+CREATE SEQUENCE sbtc_signer.canonical_chains_run_id_seq;
+
+-- Function to materialize the canonical chains for a given bitcoin chain tip
+-- and the maximum depth of the bitcoin blockchain to consider.
+--
+-- This function returns the number of rows written to the `canonical_chains`
+-- table upon success, and `-1` if rows already exist for the given chain tip.
+CREATE OR REPLACE FUNCTION sbtc_signer.materialize_canonical_chains(chain_tip BYTEA, max_depth INT)
+RETURNS INTEGER AS $$
+DECLARE
+    rows_written INTEGER;
+	existing_rows INTEGER;
+	new_run_id BIGINT;
+BEGIN
+	-- Get the next value for the run_id
+    new_run_id := nextval('sbtc_signer.canonical_chains_run_id_seq');
+
+	-- Check if rows exist with the given chain_tip
+    SELECT COUNT(*) INTO existing_rows
+    FROM sbtc_signer.canonical_chains
+    WHERE bitcoin_chain_tip = chain_tip;
+
+    -- If rows exist, return an error code
+    IF existing_rows > 0 THEN
+        RETURN -1; -- Error code indicating rows already exist
+    END IF;
+
+	-- Materialize the canonical chains from the given bitcoin chain tip.
+    WITH RECURSIVE 
+    bitcoin AS (
+        SELECT 
+            block_hash,
+            parent_hash,
+            block_height,
+			1 as depth
+        FROM sbtc_signer.bitcoin_blocks
+        WHERE block_hash = chain_tip
+
+        UNION ALL
+
+        SELECT 
+            parent.block_hash,
+            parent.parent_hash,
+            parent.block_height,
+			last.depth + 1
+        FROM sbtc_signer.bitcoin_blocks parent
+        JOIN bitcoin last ON parent.block_hash = last.parent_hash
+		WHERE last.depth < max_depth
```

Comment:
I'm actually hoping to eventually kill those functions in favor of only using the materialized data (they have horrible performance when used in joins because they're a "black box" for the query optimizer), in which case I'd prefer to keep the logic for creating them consolidated in one place.

Reply:
I'm actually hoping to eventually kill those functions in favor of only using the materialized data (they have horrible performance when used in joins because they're a "black box" for the query optimizer), in which case I'd prefer to keep the logic for creating them consolidated in one place.

Reply:
Oh okay, makes sense. I did not know that they were a black box for the query optimizer.

Reply:
Yeah, a little more info [here](https://www.postgresql.org/docs/16/xfunc-optimization.html) and [here](https://www.postgresql.org/docs/16/xfunc-volatility.html) -- we can't use `IMMUTABLE` and I did try `STABLE` flags on the functions but it didn't help :(